client/consensus/slots/src/lib.rs

```diff
@@ -319,23 +319,24 @@ pub trait SimpleSlotWorker<B: BlockT> {
 			return None
 		}
 
+		let timestamp = slot_info.duration.as_millis().saturating_mul(u64::from(slot) as u128);
```

Comment:
This is not the same as before. I mean I get the general thing you are trying to achieve. However, I would probably do it the other way around and delete the `Slot` and calculate this on the fly.

Reply:
Yeah. The difference here is that we are loosing the milliseconds part.

---

To justify my choice:
1. Timestamp here is used to be printed in debug logs and for telemetry. I thought it was not  really relevant.
2. I think of `SlotInfo` as a support structure to strictly express something about `slot`s. How it has been constructed is not relevant.
3. I wanted to detach the `slot` value from `timestamp`. There may be algorithms in the future where the slot is not computed from the timestamp (e.g. if slot is a Lamport clock or computed as described in the original Babe's paper via a median algorithm https://research.web3.foundation/en/latest/polkadot/block-production/Babe.html#-4.-clock-adjustment--relative-time-algorithm-).

---

I don't like the bounding that there is right now between `timestamp` and `slot`. For example for me the slot `InherentDataExt` [here](https://github.com/paritytech/substrate/blob/51f60d00db606e508f76e2b5f67f5eebe8647f2e/client/consensus/slots/src/lib.rs#L445-L452) should only provide a `slot()` method in it's interface.

Because of point 2 above, I would have done the opposite of what has been done in https://github.com/paritytech/substrate/pull/10906:

Instead of trying to change:
 `impl_inherent_data_provider_ext_tuple!(T, S);` to `impl_inherent_data_provider_ext_tuple!(T);

I prefer:
 `impl_inherent_data_provider_ext_tuple!(T, S);` to `impl_inherent_data_provider_ext_tuple!(S);

Again, because this is a slot information provider this should provide information about slots and not timestamp information.
How slot value is computed IMO is not relevant and should be hidden (because of 3).

---

So in the end it is only a matter about deciding: do we really require the timestamp milliseconds in the methods/functions that manipulate slots?

- If YES then we should maintain BOTH (so this PR should be closed)
- If NO then I think we have to remove the timestamp.